### PR TITLE
svm: improve integration test framework for SIMD83

### PR DESCRIPTION
**Pull request by 2501babe**
#### Problem
the first simd83 pr draft (#3146) has a nearly 2500 line diff. since the new tests im adding for transaction batches with reader-writer and writer-writer account overlap require some changes and additions to the svm integration test framework, im pulling as much as i can out into a separate pr to go in first

#### Summary of Changes
* add a new example program, `write-to-account`, which can read, write, deallocate, and reallocate an account
* add convenience helpers for this new program to the integration test file
* add a test confirming that nonce fee-payers must be rent exempt, to complement upcoming tests for what happens when a regular rent-paying fee-payer is deallocated as a side effect of transaction execution
* add compute budget and program size helper, for upcoming tests for what happens when an account is reallocated mid-batch
* improve some ergonomics, eg not advancing expected nonce state for discarded transactions, and checking transaction execution status before account states to aid test-driven development

### CODE REVIEW

File: svm/tests/example-programs/write-to-account/write_to_account_program.so
Comment (apfitzge) [inline svm/tests/example-programs/write-to-account/write_to_account_program.so:1]: never a fan of adding binary objects to the repo, but it seems this is the common practice for svm tests

File: svm/tests/example-programs/write-to-account/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+fn process_instruction(
+    _program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    data: &[u8],
+) -> ProgramResult {
+    let accounts_iter = &mut accounts.iter();
+    let target_account_info = next_account_info(accounts_iter)?;
+    match data[0] {
+        // print account size
+        0 => {
```
Comment (KirillLykov) [inline svm/tests/example-programs/write-to-account/src/lib.rs:21]: why don't we use enum here with meaningful names for variants?

File: svm/tests/example-programs/write-to-account/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                .checked_add(**target_lamports)
+                .ok_or(ProgramError::ArithmeticOverflow)?;
+
+            **target_lamports = target_lamports
+                .checked_sub(**target_lamports)
+                .ok_or(ProgramError::InsufficientFunds)?;
+        }
+        // reallocate account
+        3 => {
+            let new_size = usize::from_le_bytes(data[1..9].try_into().unwrap());
```
Comment (KirillLykov) [inline svm/tests/example-programs/write-to-account/src/lib.rs:52]: Do you avoid using borsh  to avoid associated deserialization cost?

File: svm/tests/example-programs/write-to-account/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    _program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    data: &[u8],
+) -> ProgramResult {
+    let accounts_iter = &mut accounts.iter();
+    let target_account_info = next_account_info(accounts_iter)?;
+    match data[0] {
+        // print account size
+        0 => {
+            msg!(
```
Comment (KirillLykov) [inline svm/tests/example-programs/write-to-account/src/lib.rs:22]: Do you print here for debug purpose or to imitate reading the data without writing back? In case of the second, I believe `msg` cost is not negligible and instead I would call `std::ptr::read_volatile` which prevents compiler to optimize unused code (which doesn't happen with our compiler anyways).

File: svm/tests/example-programs/write-to-account/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    program_error::ProgramError,
+    pubkey::Pubkey,
+};
+
+entrypoint!(process_instruction);
+
+fn process_instruction(
+    _program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    data: &[u8],
```
Comment (KirillLykov) [inline svm/tests/example-programs/write-to-account/src/lib.rs:15]: Does `data` contains some random number to prevent duplicate txs? This is not the only way but the simplest

File: svm/tests/example-programs/write-to-account/src/lib.rs (hunk 1 above)
Comment (2501babe) [reply svm/tests/example-programs/write-to-account/src/lib.rs:21]: like the other example programs that already exist, i just wanted to do the simplest thing possible. theyre only used for one set of tests and arent expected to change, and are trivial 20-40 line single-file programs that arent part of the toplevel monorepo `Cargo.toml` workspace and arent imported by anything. i have helpers in `svm/tests/integration_test.rs` tho

File: svm/tests/example-programs/write-to-account/src/lib.rs (hunk 2 above)
Comment (2501babe) [reply svm/tests/example-programs/write-to-account/src/lib.rs:52]: cost isnt a concern, as above its just doing the simplest thing possible

File: svm/tests/example-programs/write-to-account/src/lib.rs (hunk 3 above)
Comment (2501babe) [reply svm/tests/example-programs/write-to-account/src/lib.rs:22]: i use this to check which specific transactions cause accounts to be deallocated or reallocated in a multi-transaction batch. svm integration tests can directly read transaction logs off the execution result

File: svm/tests/mock_bank.rs [hunk 5]
```diff
@@ -138,6 +138,11 @@ pub fn program_address(program_name: &str) -> Pubkey {
     Pubkey::create_with_seed(&Pubkey::default(), program_name, &Pubkey::default()).unwrap()
 }
 
+#[allow(unused)]
+pub fn program_data_size(program_name: &str) -> usize {
+    load_program(program_name.to_string()).len()
+}
```
Comment (apfitzge) [inline svm/tests/mock_bank.rs:144]: what is this for?

File: svm/tests/example-programs/write-to-account/src/lib.rs (hunk 4 above)
Comment (apfitzge) [reply svm/tests/example-programs/write-to-account/src/lib.rs:15]: no check on length - if need that functionality easy to just append random bytes at the end; program has no need to read them

Comment (2501babe) [reply svm/tests/example-programs/write-to-account/src/lib.rs:15]: no, duplicates arent a concern because all the svm integration tests are structured like "here is a specific list of transactions to execute, heres what the execution results, final account states, and log messages should be"

File: svm/tests/integration_test.rs [hunk 6]
```diff
@@ -682,8 +706,12 @@ fn simple_nonce(enable_fee_only_transactions: bool, fee_paying_nonce: bool) -> V
             let mut fee_payer_data = AccountSharedData::default();
             fee_payer_data.set_lamports(LAMPORTS_PER_SOL);
             test_entry.add_initial_account(fee_payer, &fee_payer_data);
+        } else if rent_paying_nonce {
+            assert!(fee_paying_nonce);
```
Comment (apfitzge) [inline svm/tests/integration_test.rs:710]: not sure I understand the connection between fee paying and rent paying for the nonce here.

File: svm/tests/integration_test.rs [hunk 7]
```diff
@@ -841,9 +870,105 @@ fn simple_nonce(enable_fee_only_transactions: bool, fee_paying_nonce: bool) -> V
...
+
+        test_entry.push_nonce_transaction_with_status(
+            transaction,
+            nonce_info.clone(),
+            ExecutionStatus::Discarded,
+        );
+    }
+
+    // 5: rent-paying nonce fee-payers are also not charged for fee-only transactions
+    if enable_fee_only_transactions && fee_paying_nonce {
```
Comment (apfitzge) [inline svm/tests/integration_test.rs:895]: this seems exactly the same as above, or am I missing something?

**Review by apfitzge — COMMENTED** [review]
Changes look fine to me; there seem to be a few unrelated changes here which made reviewing it a bit confusing tbh, but not going to block on it.

Commented a few questions

**Comment by 2501babe** [comment]
sorry this is a bit confusing out of context, the goal is purely to make reviewing #3146 easier (since it is extremely large) by a) reducing it by a few hundred loc, and b) making its diff of `integration_test.rs` purely adding more tests, with no changes to the code that runs tests

File: svm/tests/mock_bank.rs (hunk 5 above)
Comment (2501babe) [reply svm/tests/mock_bank.rs:144]: a very tricky edge case test in #3146 where i reallocate an account to be much larger and set a transaction data size limit that is larger than the program size plus old account size, but smaller than the program size plus new account size, to ensure that the transaction data size calculation uses the reallocated size (this affects consensus)

File: svm/tests/integration_test.rs (hunk 6 above)
Comment (2501babe) [reply svm/tests/integration_test.rs:710]: a nonce account used as a transaction fee-payer that would be brought one lamport below rent exemption if it was allowed to pay for the transaction

File: svm/tests/integration_test.rs (hunk 7 above)
Comment (2501babe) [reply svm/tests/integration_test.rs:895]: maybe im overly paranoid but case 4 uses a real program id, so account loading can succeed, but the transaction is discarded because charging for it would make the valid nonce fee-payer become invalid. case 5 uses a fake program id, which aborts account loading, so we test that with fee-only transactions enabled there isnt any kind of short-circuit code path that would cause it to pay fees

i wanted to be very thorough about this because there are a number of edge cases to cover if a) a regular fee-payer has to pay rent, and b) nonce accounts are mutated in-batch, so we want to be absolutely sure the overlap of those two things is impossible

File: svm/tests/integration_test.rs (hunk 6 above)
Comment (2501babe) [reply svm/tests/integration_test.rs:710]: oh you mean the assert, its just to make sure the annoyingly complex lambda wasnt misused. if you wanted a rent-paying non-fee-paying nonce you wouldnt add `LAMPORTS_PER_SIGNATURE`, or else youd still have a rent-exempt nonce, and might write a test that doesnt test what it means to test. but no tests need an account like that so i didnt add a branch for it

File: svm/tests/integration_test.rs (hunk 7 above)
Comment (apfitzge) [reply svm/tests/integration_test.rs:895]: thanks. I just didn't notice the difference in 2nd arge of `mk_nonce_transaction`

File: svm/tests/integration_test.rs [hunk 8]
```diff
@@ -841,9 +870,105 @@ fn simple_nonce(enable_fee_only_transactions: bool, fee_paying_nonce: bool) -> V
...
+        };
+
+        let account_metas = match self {
+            Self::Dealloc => vec![
+                AccountMeta::new(target, false),
+                AccountMeta::new(solana_sdk::incinerator::id(), false),
+            ],
+            Self::Print => vec![AccountMeta::new_readonly(target, false)],
+            _ => vec![AccountMeta::new(target, false)],
+        };
```
Comment (apfitzge) [inline svm/tests/integration_test.rs]: nit: can we just couple these together so that there's 1 match and it's very clear what the ix_data and account inputs are for each instruction variant?

**Review by apfitzge — APPROVED** [review]
lgtm
